docs/developer/best_practices/commit.rst

```diff
@@ -0,0 +1,356 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Why do we need commit message
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+After comments in the code, commit messages are the easiest way to find context
+for every single line of code: running ``git blame`` on a file will give you,
+for each line, the identifier of the last commit that changed the line.
+
+Unlike a comment in the code (which applies to a single line or file), a commit
+message applies to a logical change and thus can provide information on the
+design of the code and why the change was done. This makes commit messages a
+part of the code documentation and makes them helpful for other developers to
+understand your code.
+
+Last but not least: commit messages can also be used for automating tasks such
+as issue management and changelog generation.
+
+Note that it is important to have all the necessary information in the commit
+message, instead of having them (only) in the related issue, because:
+
+- the issue can contain troubleshooting/design discussion/investigation with a
+  lot of back and forth, which makes hard to get the gist of it.
+
+- you need access to an external service to get the whole context, which goes
+  against one of biggest advantage of the distributed SCM (having all the
+  information you need offline, from your local copy of the repository).
+
+Anatomy of a good commit message
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A commit is composed of a subject, a body and a footer. A blank line separates
+the subject from body and the body from the footer.
+
+The body can be omitted for trivial commit. That being said, be very careful:
+a change might seem trivial when you write it but will seem totally awkward
+the day you will have to understand why you made it. If you think your patch is
+trivial and somebody tells you he does not understand your patch, then your
+patch is not trivial and it requires a detailed description.
+
+The footer contains references for issue management (``Refs``, ``Closes``, …)
+or other relevant annotations (cherry-pick source, …).
+Optional if your commit is not related to any issue (should be pretty rare).
+
+Subject
+:::::::
+
+A good commit message should start with a short summary of the change: the
+subject line.
+
+This summary should be written using the imperative mood and carry as much
+information as possible while staying short, ideally under 50 characters (this
+is a goal, the hard limit is 72).
+
+Do not end the subject line with a period.
+
+It is composed of:
+
+- a topic, usually the name of the affected component (``ui``, ``build``,
+  ``docs``, …)
+
+- a slash and then the name of the sub-component (optional)
+
+- a colon
+
+- the description of the change
+
+Examples:
+
+- ``ci: use proxy-cache to reduce flakiness``
+
+- ``build/package: factorize task_dep in DEBPackage``
+
+- ``ui/volume: add banner when failed to create volume``
+
+If several components are affected:
+
+- split your commit (preferred)
+
+- pick only the most affected one
```

Comment:
or comma-separate?

Reply:
It's an option, but it will nibble at the already tight character budget of the subject line.

So I'm not sure it's a good one :/

Reply:
I'd even argue those *topics* in a commit subject are already expensive to use: e.g. `examples/solution: ` costs 19 out of 50 characters, which makes meaningful subjects harder to write.

I'm wondering what the use-case of *topics* really is. Are we trying to have easy to understand lists of commits, from `git log --oneline` and the likes? Or to have easy to filter lists of commits? If the latter, then having such information in the subject is unnecessary.

Reply:
Yeah, topic can be quite a mouthful as well, but in you example I would replace `examples/solution` by `solution` since it stays unambiguous.

The "heuristic" I give (take the name of the top-level directory) is not perfect, it's more like an hint that works is most cases, but is not always optimal.

Reply:
There's a slight difference between `examples/solution` and `salt/solutions` though. But yes, let's do our best :)

Reply:
Yeah, we could interpret `solutions` as the `solutions` itself, and `salt/solutions` as the Salt code to deploy the solution.
But yeah, no hard rules (that's why I haven't listed a closed-set of accepted topic: the goal is mainly informative).